File: b680(track_arrangement)/Python/track_arrangement.py
from sys import stdin, stdout
from functools import cmp_to_key

def arrange_team(runner_1, runner_2):
    return runner_1[1] - runner_2[1]

def is_even(n):
    if n % 2: return False
    return True

while True:
    num_of_runners = stdin.readline().strip('\n')
    if not num_of_runners: break
    num_of_runners = int(num_of_runners)
    runners = []
    for n in range(num_of_runners):
        runners.append(tuple(map(float, stdin.readline().split())))
    runners.sort(key = cmp_to_key(arrange_team))
    teams = [[] for i in range(num_of_runners // 8)]
    if num_of_runners % 8: teams.append([])
    num_of_teams = len(teams)
    for rank, runner in enumerate(runners):
        if is_even(rank // num_of_teams):
            teams[rank % num_of_teams].append(runner)
        else:
            teams[num_of_teams - 1 - rank % num_of_teams].append(runner)
    for team_num in range(num_of_teams - 1):
        print(team_num+1, end = ' ')
        team = teams[team_num]
        print(int(team[6][0]), int(team[4][0]), int(team[2][0]), int(team[0][0]), int(team[1][0]), int(team[3][0]), int(team[5][0]), int(team[7][0]))
    team = teams[num_of_teams-1]
    print(num_of_teams, end = ' ')
    s = len(team)
    if is_even(s):
        for i in range(s-2, -1, -2):
            print(int(team[i][0]), end = ' ')
        for i in range(1, s, 2):
            print(int(team[i][0]), end = ' ')
    else:
        for i in range(s-1, -1, -2):
            print(int(team[i][0]), end = ' ')
        for i in range(1, s, 2):
            print(int(team[i][0]), end = ' ')
    print()

File: b680(track_arrangement)/C++/track_arrangement.cpp
#include <iostream>
#include <vector>
#include <utility>
#include <algorithm>

using namespace std;

inline bool is_even(int n) {
  return n % 2? false: true;
}

int main(int argc, char const *argv[]) {
  ios_base::sync_with_stdio(false);
  cin.tie(0);
  int num_of_runners;
  while (cin >> num_of_runners) {
    int runner_num;
    double score;
    vector<pair<int, double> > runners;
    for (int i = 0; i < num_of_runners; i++) {
      cin >> runner_num >> score;
      runners.push_back(pair<int, double>(runner_num, score));
    }
    sort(runners.begin(), runners.end(), [](pair<int, double> runner_1, pair<int, double> runner_2) {
      return runner_1.second < runner_2.second;
    });
    int num_of_teams = num_of_runners%8?num_of_runners/8+1:num_of_runners/8;
    vector<vector<pair<int, double> > > teams(num_of_teams);
    for (int i = 0; i < num_of_runners; i++) {
      if (is_even(i / num_of_teams)) teams.at(i % num_of_teams).push_back(runners.at(i));
      else teams.at(num_of_teams - 1 - i % num_of_teams).push_back(runners.at(i));
    }
    for (int i = 0; i < num_of_teams - 1; i++) {
      vector<pair<int, double> > team = teams.at(i);
      cout << i+1 << ' ' << team.at(6).first << ' ' << team.at(4).first << ' ' << team.at(2).first << ' ' << team.at(0).first << ' ' << team.at(1).first << ' ' << team.at(3).first << ' ' << team.at(5).first << ' ' << team.at(7).first << '\n';
    }
    vector<pair<int, double> > team = teams.at(num_of_teams - 1);
    cout << num_of_teams << ' ';
    int s = team.size();
    if (is_even(s)) {
      for (int i = s - 2; i >= 0; i -= 2) std::cout << team.at(i).first << ' ';
      for (int i = 1; i < s; i += 2) std::cout << team.at(i).first << ' ';
    } else {
      for (int i = s - 1; i >= 0; i -= 2) std::cout << team.at(i).first << ' ';
      for (int i = 1; i < s; i += 2) std::cout << team.at(i).first << ' ';
    }
    std::cout << '\n';
  }
  return 0;
}
